skfp/model_selection/splitters/pubchem_split.py

```diff
@@ -0,0 +1,456 @@
+) -> list[list[int]]:
+    """
+    Get first literature publication year from PubChem for a list of molecules, either
+    as SMILES strings or RDKit Mol objects.
+
+    Parameters
+    ----------
+    data : sequence
+        A sequence representing either SMILES strings or RDKit `Mol` objects.
+    """
+    data = list(
+        map(
+            lambda mol: (
+                MolToSmiles(mol, canonical=True, isomericSmiles=True)
+                if mol is Mol
+                else mol
+            ),
+            data,
+        )
+    )
+
+    print("Converting SMILES to CIDs")
+    cids = run_in_parallel(get_cid_for_smiles, data, n_jobs)
+
+    print("Searching for CIDs literature")
+    years = run_in_parallel(get_earliest_publication_date, cids, n_jobs)
+
+    return years
+
+
+def get_cid_for_smiles(smiles: str) -> Optional[str]:
+    """
+
+    Parameters:
+    ------------
+    smiles: str
+        SMILES string
+
+    Return
+    --------
+        PubChem compound identifier if  else None
+
+    """
+    PUG_url = f"https://pubchem.ncbi.nlm.nih.gov/rest/pug/compound/smiles/{quote(smiles)}/cids/JSON"
+
+    PUG_response = requests.get(PUG_url, timeout=10)
+    PUG_response_status = PUG_response.status_code
+
+    if PUG_response_status != 200:
+        warnings.warn(
+            f"PUG error \n, {PUG_url} \n status code {PUG_response_status}, \n check your input"
+        )
+
+    PUG_response_json = PUG_response.json()
+
+    if "IdentifierList" in PUG_response_json:
+        cid = PUG_response_json["IdentifierList"]["CID"][0]
+        if bool(cid):
+            cid = str(cid)
+        else:
+            cid = None
+    else:
+        cid = None
+    return cid
+
+
+def get_earliest_publication_date(cid: Union[str, None]) -> Optional[int]:
+    """
+    Get the date of the earliest publication from PubChem where a given molecule appears.
+    There are molecules without publications, for which we return None.
+
+    Parameters
+    ----------
+    Optional[int]
+        year of earliest publication date associated to cid
+
+    Returns
+    ----------
+
+    """
+    if not cid:
+        return None
+
+    query = {
+        "download": "*",
+        "collection": "literature",
+        "where": {"ands": [{"cid": str(cid)}]},
+        "order": ["articlepubdate,asc"],
+        "start": 0,
+        "limit": 3,  # fetch a few, just in case
+        "downloadfilename": f"pubchem_cid_{cid}_literature",
+        "nullatbottom": 1,
+    }
+    base_url = "https://pubchem.ncbi.nlm.nih.gov/sdq/sdqagent.cgi"
+    params = {"infmt": "json", "outfmt": "json", "query": json.dumps(query)}
+
+    def get_publication_date(response: list[dict]) -> Optional[int]:
+        article_pub_date = None
+        for entry in response:
+            if "articlepubdate" in entry:
+                return int(entry["articlepubdate"][:4])  # get year of publication
+            else:
+                continue
+        return article_pub_date
+
+    while True:
```

Comment:
This should be limited, like the function above, with configurable num retries and timeout